server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java

```diff
@@ -190,8 +202,19 @@ private void fillParallelArray(ScoreDoc[] scoreDocs, ParallelArray parallelArray
                 parallelArray.hasRecoverySource[index] = combinedDocValues.hasRecoverySource(segmentDocID);
             }
             // now sort back based on the shardIndex. we use this to store the previous index
-            ArrayUtil.introSort(scoreDocs, Comparator.comparingInt(i -> i.shardIndex));
+            if (parallelArray.useSequentialStoredFieldsReader == false) {
+                ArrayUtil.introSort(scoreDocs, Comparator.comparingInt(i -> i.shardIndex));
+            }
+        }
+    }
+
+    private static boolean hasSequentialAccess(ScoreDoc[] scoreDocs) {
```

Comment:
You can avoid the loop entirely like in [FetchPhase#hasSequentialDocs](https://github.com/elastic/elasticsearch/blob/780f27306777399db574b9f9a9f44e77f41779f3/server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java#L446)

Reply:
This is a bit different, where we can access documents out of order instead of ascending like in the fetch phase.